datafusion/optimizer/src/eliminate_join.rs

```diff
@@ -0,0 +1,122 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::optimizer::ApplyOrder;
+use crate::{OptimizerConfig, OptimizerRule};
+use datafusion_common::{Result, ScalarValue};
+use datafusion_expr::JoinType::Inner;
+use datafusion_expr::{
+    logical_plan::{EmptyRelation, LogicalPlan},
+    CrossJoin, Expr,
+};
+
+/// Eliminates joins when inner join condition is false.
+/// Replaces joins when inner join condition is true with a cross join.
+#[derive(Default)]
+pub struct EliminateJoin;
+
+impl EliminateJoin {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for EliminateJoin {
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        match plan {
+            LogicalPlan::Join(join) if join.join_type == Inner && join.on.is_empty() => {
+                match join.filter {
+                    Some(Expr::Literal(ScalarValue::Boolean(Some(v)))) => {
```

Comment:
I don't know if it would be better, but you could also write this using two match arms:

```
Some(Expr::Literal(ScalarValue::Boolean(Some(true)))) => {
...
}
Some(Expr::Literal(ScalarValue::Boolean(Some(false)))) => {
...
}
```

Reply:
It's a nice improvement